File: python/doppelganger_py/expressions.py
"""Expression evaluation system for matching request parameters."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any, Callable
import re


@dataclass
class EvaluationFetchers:
    """Container for functions that fetch values from the request."""
    body_fetcher: dict[str, Any]
    query_fetcher: Callable[[str], str]
    query_array_fetcher: Callable[[str], list[str]]
    param_fetcher: Callable[[str], str]


class Expression(ABC):
    """Base class for all expressions."""
    
    @abstractmethod
    def evaluate(self, fetchers: EvaluationFetchers) -> Any:
        """Evaluate the expression and return a result."""
        pass
    
    @abstractmethod
    def return_type(self) -> type:
        """Return the type of the evaluation result."""
        pass


class AndExpression(Expression):
    """Logical AND of multiple boolean expressions."""
    
    def __init__(self, expressions: list[Expression]):
        self.expressions = expressions
    
    def evaluate(self, fetchers: EvaluationFetchers) -> bool:
        for expr in self.expressions:
            if not expr.evaluate(fetchers):
                return False
        return True
    
    def return_type(self) -> type:
        return bool


class OrExpression(Expression):
    """Logical OR of multiple boolean expressions."""
    
    def __init__(self, expressions: list[Expression]):
        self.expressions = expressions
    
    def evaluate(self, fetchers: EvaluationFetchers) -> bool:
        for expr in self.expressions:
            if expr.evaluate(fetchers):
                return True
        return False
    
    def return_type(self) -> type:
        return bool


class NotExpression(Expression):
    """Logical NOT of a boolean expression."""
    
    def __init__(self, expression: Expression):
        self.expression = expression
    
    def evaluate(self, fetchers: EvaluationFetchers) -> bool:
        return not self.expression.evaluate(fetchers)
    
    def return_type(self) -> type:
        return bool


class ContainsExpression(Expression):
    """Check if a list contains all specified values."""
    
    def __init__(self, list_expr: Expression, values: list[Expression]):
        self.list_expr = list_expr
        self.values = values
    
    def evaluate(self, fetchers: EvaluationFetchers) -> bool:
        list_values = self.list_expr.evaluate(fetchers)
        for value_expr in self.values:
            if value_expr.evaluate(fetchers) not in list_values:
                return False
        return True
    
    def return_type(self) -> type:
        return bool


class EqualsExpression(Expression):
    """Check if two expressions are equal."""
    
    def __init__(self, left: Expression, right: Expression):
        self.left = left
        self.right = right
    
    def evaluate(self, fetchers: EvaluationFetchers) -> bool:
        left_val = self.left.evaluate(fetchers)
        right_val = self.right.evaluate(fetchers)
        return left_val == right_val
    
    def return_type(self) -> type:
        return bool


class RegexExpression(Expression):
    """Check if a value matches a regex pattern."""
    
    def __init__(self, value: Expression, pattern: str):
        self.value = value
        self.pattern = pattern
        self._compiled = re.compile(pattern)
    
    def evaluate(self, fetchers: EvaluationFetchers) -> bool:
        value = self.value.evaluate(fetchers)
        return bool(self._compiled.search(value))
    
    def return_type(self) -> type:
        return bool


class BodyValueExpression(Expression):
    """Get a value from the request body."""
    
    def __init__(self, id: str):
        self.id = id
    
    def evaluate(self, fetchers: EvaluationFetchers) -> str:
        value = fetchers.body_fetcher.get(self.id)
        return str(value) if value is not None else ""
    
    def return_type(self) -> type:
        return str


class QueryValueExpression(Expression):
    """Get a value from query parameters."""
    
    def __init__(self, id: str):
        self.id = id
    
    def evaluate(self, fetchers: EvaluationFetchers) -> str:
        return fetchers.query_fetcher(self.id)
    
    def return_type(self) -> type:
        return str


class QueryArrayValueExpression(Expression):
    """Get an array value from query parameters."""
    
    def __init__(self, id: str):
        self.id = id
    
    def evaluate(self, fetchers: EvaluationFetchers) -> list[str]:
        value = fetchers.query_fetcher(self.id)
        if "," in value:
            return value.split(",")
        return fetchers.query_array_fetcher(self.id)
    
    def return_type(self) -> type:
        return list


class PathValueExpression(Expression):
    """Get a value from path parameters."""
    
    def __init__(self, id: str):
        self.id = id
    
    def evaluate(self, fetchers: EvaluationFetchers) -> str:
        return fetchers.param_fetcher(self.id)
    
    def return_type(self) -> type:
        return str


class StringValueExpression(Expression):
    """A literal string value."""
    
    def __init__(self, value: str):
        self.value = value
    
    def evaluate(self, fetchers: EvaluationFetchers) -> str:
        return self.value
    
    def return_type(self) -> type:
        return str


# Expression factory registry
EXPRESSION_REGISTRY: dict[str, Callable[[dict], Expression]] = {}


def register_expression(type_name: str):
    """Decorator to register an expression factory."""
    def decorator(factory: Callable[[dict], Expression]):
        EXPRESSION_REGISTRY[type_name] = factory
        return factory
    return decorator


@register_expression("AND")
def and_factory(data: dict) -> Expression:
    expressions = [build_expression(expr) for expr in data["expressions"]]
    for expr in expressions:
        if expr.return_type() != bool:
            raise ValueError("AND values must be bool")
    return AndExpression(expressions)


@register_expression("OR")
def or_factory(data: dict) -> Expression:
    expressions = [build_expression(expr) for expr in data["expressions"]]
    for expr in expressions:
        if expr.return_type() != bool:
            raise ValueError("OR values must be bool")
    return OrExpression(expressions)


@register_expression("NOT")
def not_factory(data: dict) -> Expression:
    expression = build_expression(data["expression"])
    if expression.return_type() != bool:
        raise ValueError("NOT value must be bool")
    return NotExpression(expression)


@register_expression("CONTAINS")
def contains_factory(data: dict) -> Expression:
    if "values" not in data:
        raise ValueError("CONTAINS must have values attribute")
    if "list" not in data:
        raise ValueError("CONTAINS must have list attribute")
    
    values = [build_expression(v) for v in data["values"]]
    for v in values:
        if v.return_type() != str:
            raise ValueError("CONTAINS values must be string")
    
    list_expr = build_expression(data["list"])
    if list_expr.return_type() != list:
        raise ValueError("CONTAINS list must be list")
    
    return ContainsExpression(list_expr, values)


@register_expression("EQUALS")
def equals_factory(data: dict) -> Expression:
    right = build_expression(data["right"])
    left = build_expression(data["left"])
    
    if right.return_type() != left.return_type():
        raise ValueError("EQUALS right and left must be the same type")
    
    return EqualsExpression(left, right)


@register_expression("REGEX")
def regex_factory(data: dict) -> Expression:
    value = build_expression(data["value"])
    pattern = data["pattern"]
    
    if value.return_type() != str:
        raise ValueError("REGEX value must be string")
    
    return RegexExpression(value, pattern)


@register_expression("BODY")
def body_factory(data: dict) -> Expression:
    return BodyValueExpression(data["id"])


@register_expression("QUERY")
def query_factory(data: dict) -> Expression:
    return QueryValueExpression(data["id"])


@register_expression("QUERY_ARRAY")
def query_array_factory(data: dict) -> Expression:
    return QueryArrayValueExpression(data["id"])


@register_expression("PATH")
def path_factory(data: dict) -> Expression:
    return PathValueExpression(data["id"])


@register_expression("STRING")
def string_factory(data: dict) -> Expression:
    return StringValueExpression(data["value"])


def build_expression(data: dict) -> Expression:
    """Build an expression from a dictionary."""
    type_name = data.get("type")
    if type_name not in EXPRESSION_REGISTRY:
        raise ValueError(f"Unknown expression type: {type_name}")
    
    factory = EXPRESSION_REGISTRY[type_name]
    return factory(data)

File: typescript/src/expressions.ts
/**
 * Expression evaluation system for matching request parameters.
 */

// Type for fetcher functions
export interface EvaluationFetchers {
  bodyFetcher: Record<string, unknown>;
  queryFetcher: (key: string) => string;
  queryArrayFetcher: (key: string) => string[];
  paramFetcher: (key: string) => string;
}

// Return type enum
export type ReturnType = 'boolean' | 'string' | 'array';

// Base expression interface
export interface Expression {
  evaluate(fetchers: EvaluationFetchers): unknown;
  returnType(): ReturnType;
}

// Expression data from JSON
export interface ExpressionData {
  type: string;
  [key: string]: unknown;
}

// Expression factory type
type ExpressionFactory = (data: ExpressionData) => Expression;

// Expression registry
const expressionRegistry: Map<string, ExpressionFactory> = new Map();

/**
 * Register an expression factory
 */
function registerExpression(typeName: string, factory: ExpressionFactory): void {
  expressionRegistry.set(typeName, factory);
}

/**
 * AND Expression - All expressions must be true
 */
class AndExpression implements Expression {
  constructor(private readonly expressions: Expression[]) {}

  evaluate(fetchers: EvaluationFetchers): boolean {
    for (const expr of this.expressions) {
      if (!expr.evaluate(fetchers)) {
        return false;
      }
    }
    return true;
  }

  returnType(): ReturnType {
    return 'boolean';
  }
}

/**
 * OR Expression - At least one expression must be true
 */
class OrExpression implements Expression {
  constructor(private readonly expressions: Expression[]) {}

  evaluate(fetchers: EvaluationFetchers): boolean {
    for (const expr of this.expressions) {
      if (expr.evaluate(fetchers)) {
        return true;
      }
    }
    return false;
  }

  returnType(): ReturnType {
    return 'boolean';
  }
}

/**
 * NOT Expression - Negates a boolean expression
 */
class NotExpression implements Expression {
  constructor(private readonly expression: Expression) {}

  evaluate(fetchers: EvaluationFetchers): boolean {
    return !this.expression.evaluate(fetchers);
  }

  returnType(): ReturnType {
    return 'boolean';
  }
}

/**
 * CONTAINS Expression - Check if list contains all specified values
 */
class ContainsExpression implements Expression {
  constructor(
    private readonly listExpr: Expression,
    private readonly values: Expression[]
  ) {}

  evaluate(fetchers: EvaluationFetchers): boolean {
    const listValues = this.listExpr.evaluate(fetchers) as string[];
    for (const valueExpr of this.values) {
      const value = valueExpr.evaluate(fetchers) as string;
      if (!listValues.includes(value)) {
        return false;
      }
    }
    return true;
  }

  returnType(): ReturnType {
    return 'boolean';
  }
}

/**
 * EQUALS Expression - Compare two values for equality
 */
class EqualsExpression implements Expression {
  constructor(
    private readonly left: Expression,
    private readonly right: Expression
  ) {}

  evaluate(fetchers: EvaluationFetchers): boolean {
    const leftVal = this.left.evaluate(fetchers);
    const rightVal = this.right.evaluate(fetchers);

    if (Array.isArray(leftVal) && Array.isArray(rightVal)) {
      if (leftVal.length !== rightVal.length) {
        return false;
      }
      return leftVal.every((val, idx) => val === rightVal[idx]);
    }

    return leftVal === rightVal;
  }

  returnType(): ReturnType {
    return 'boolean';
  }
}

/**
 * REGEX Expression - Match a value against a regex pattern
 */
class RegexExpression implements Expression {
  private readonly regex: RegExp;

  constructor(
    private readonly value: Expression,
    pattern: string
  ) {
    this.regex = new RegExp(pattern);
  }

  evaluate(fetchers: EvaluationFetchers): boolean {
    const value = this.value.evaluate(fetchers) as string;
    return this.regex.test(value);
  }

  returnType(): ReturnType {
    return 'boolean';
  }
}

/**
 * BODY Expression - Get a value from the request body
 */
class BodyValueExpression implements Expression {
  constructor(private readonly id: string) {}

  evaluate(fetchers: EvaluationFetchers): string {
    const value = fetchers.bodyFetcher[this.id];
    return value !== undefined && value !== null ? String(value) : '';
  }

  returnType(): ReturnType {
    return 'string';
  }
}

/**
 * QUERY Expression - Get a value from query parameters
 */
class QueryValueExpression implements Expression {
  constructor(private readonly id: string) {}

  evaluate(fetchers: EvaluationFetchers): string {
    return fetchers.queryFetcher(this.id);
  }

  returnType(): ReturnType {
    return 'string';
  }
}

/**
 * QUERY_ARRAY Expression - Get an array value from query parameters
 */
class QueryArrayValueExpression implements Expression {
  constructor(private readonly id: string) {}

  evaluate(fetchers: EvaluationFetchers): string[] {
    const value = fetchers.queryFetcher(this.id);
    if (value.includes(',')) {
      return value.split(',');
    }
    return fetchers.queryArrayFetcher(this.id);
  }

  returnType(): ReturnType {
    return 'array';
  }
}

/**
 * PATH Expression - Get a value from path parameters
 */
class PathValueExpression implements Expression {
  constructor(private readonly id: string) {}

  evaluate(fetchers: EvaluationFetchers): string {
    return fetchers.paramFetcher(this.id);
  }

  returnType(): ReturnType {
    return 'string';
  }
}

/**
 * STRING Expression - A literal string value
 */
class StringValueExpression implements Expression {
  constructor(private readonly value: string) {}

  evaluate(_fetchers: EvaluationFetchers): string {
    return this.value;
  }

  returnType(): ReturnType {
    return 'string';
  }
}

// Register all expression factories
registerExpression('AND', (data: ExpressionData): Expression => {
  const expressionsData = data.expressions as ExpressionData[];
  const expressions = expressionsData.map((expr) => buildExpression(expr));
  for (const expr of expressions) {
    if (expr.returnType() !== 'boolean') {
      throw new Error('AND values must be boolean');
    }
  }
  return new AndExpression(expressions);
});

registerExpression('OR', (data: ExpressionData): Expression => {
  const expressionsData = data.expressions as ExpressionData[];
  const expressions = expressionsData.map((expr) => buildExpression(expr));
  for (const expr of expressions) {
    if (expr.returnType() !== 'boolean') {
      throw new Error('OR values must be boolean');
    }
  }
  return new OrExpression(expressions);
});

registerExpression('NOT', (data: ExpressionData): Expression => {
  const expressionData = data.expression as ExpressionData;
  const expression = buildExpression(expressionData);
  if (expression.returnType() !== 'boolean') {
    throw new Error('NOT value must be boolean');
  }
  return new NotExpression(expression);
});

registerExpression('CONTAINS', (data: ExpressionData): Expression => {
  if (!data.values) {
    throw new Error('CONTAINS must have values attribute');
  }
  if (!data.list) {
    throw new Error('CONTAINS must have list attribute');
  }

  const valuesData = data.values as ExpressionData[];
  const values = valuesData.map((v) => buildExpression(v));
  for (const v of values) {
    if (v.returnType() !== 'string') {
      throw new Error('CONTAINS values must be string');
    }
  }

  const listExpr = buildExpression(data.list as ExpressionData);
  if (listExpr.returnType() !== 'array') {
    throw new Error('CONTAINS list must be array');
  }

  return new ContainsExpression(listExpr, values);
});

registerExpression('EQUALS', (data: ExpressionData): Expression => {
  const right = buildExpression(data.right as ExpressionData);
  const left = buildExpression(data.left as ExpressionData);

  if (right.returnType() !== left.returnType()) {
    throw new Error('EQUALS right and left must be the same type');
  }

  return new EqualsExpression(left, right);
});

registerExpression('REGEX', (data: ExpressionData): Expression => {
  const value = buildExpression(data.value as ExpressionData);
  const pattern = data.pattern as string;

  if (value.returnType() !== 'string') {
    throw new Error('REGEX value must be string');
  }

  return new RegexExpression(value, pattern);
});

registerExpression('BODY', (data: ExpressionData): Expression => {
  return new BodyValueExpression(data.id as string);
});

registerExpression('QUERY', (data: ExpressionData): Expression => {
  return new QueryValueExpression(data.id as string);
});

registerExpression('QUERY_ARRAY', (data: ExpressionData): Expression => {
  return new QueryArrayValueExpression(data.id as string);
});

registerExpression('PATH', (data: ExpressionData): Expression => {
  return new PathValueExpression(data.id as string);
});

registerExpression('STRING', (data: ExpressionData): Expression => {
  return new StringValueExpression(data.value as string);
});

/**
 * Build an expression from JSON data
 */
export function buildExpression(data: ExpressionData): Expression {
  const typeName = data.type;
  const factory = expressionRegistry.get(typeName);

  if (!factory) {
    throw new Error(`Unknown expression type: ${typeName}`);
  }

  return factory(data);
}
